### PR TITLE
Fix use of # character in Racket lexer

### LINKED ISSUE #1471: Racket number sign (hashtag) highlighting
I'm experiencing issues with Racket language number signs (hashtags) highlighting.
 
**Name of the lexer**
Racket.

**Code sample**
Sample 1.
```racket
#| Block comment 
started with the number sign (hashtag).
(define x 3)
should be highlighted as a commented text.
|#
```
Sample 2.
```racket
(define ht #true)
```
[Demo on Gitlab](https://gitlab.com/yugenekr/test-markdown-formatting/-/blob/master/highlight-comment-block.rkt)
[Demo on rouge.jneen.net](http://rouge.jneen.net/v3.17.0/racket/I3xCbG9jayBjb21tZW50LgooZGVmaW5lIHggMykKaXMgbm90IGhpZ2hsaWdodGVkIGFzIHRoZSBjb21tZW50ZWQgdGV4dC4gCnwjCgojfApCbG9jayBjb21tZW50LgooZGVmaW5lIHggMykKaXMgbm90IGhpZ2hsaWdodGVkIGFzIHRoZSBjb21tZW50ZWQgdGV4dC4KfCMKCjs7ICN0cnVlIGlzIG5vdCBoaWdobGlnaHRlZCBwcm9wZXJseQooZGVmaW5lIGh0ICN0cnVlKQoKCgo)

**Additional context**
I was sent here after filing [the issue on Gitlab](https://gitlab.com/gitlab-org/gitlab/-/issues/205935). I'm not aware of any specifics of Rouge library.

### PR DESCRIPTION
**Pull request by pyrmont**
The `#` character can be used in Racket in multiline comments as well as in boolean constants. As described in #1471, the Racket lexer does not at present correctly lex multiline comments or full-word (i.e. `#true` as opposed to `#t`) boolean constants.

This PR fixes those bugs (and will close #1471). It also removes extraneous code from the visual sample.

### CODE REVIEW

**Comment by pyrmont** [comment]
How's this look @yugenekr?

File: spec/visual/samples/racket [hunk 1]
```diff
@@ -2,11 +2,20 @@
...
+
 #| Multi-line comment on one line |#
 
 #| Multi-line comment on
    two lines |#
 
+#|
+#  Block comment.
+#  (define x 3)
+#  is not highlighted as the commented text.
```
Comment (yugenekr) [inline spec/visual/samples/racket]: Why `#` on these lines? I believe block comment should be identified just by starting #| and ending |#. Having additional `#` for this use case might be confusing.

File: spec/visual/samples/racket [hunk 2]
```diff
@@ -104,3 +104,4 @@
 
 ;; #true is not highlighted properly
 (define ht #true)
+(define hf #false)
```
Comment (yugenekr) [inline spec/visual/samples/racket]: Should we also add [`#T` `#F` `#t` `#f`](https://docs.racket-lang.org/reference/reader.html#%28part._parse-boolean%29) samples here?

File: spec/visual/samples/racket (hunk 1 above)
Comment (pyrmont) [reply spec/visual/samples/racket]: My bad. Didn't notice Vim had stuck those in there. Will remove.

File: spec/visual/samples/racket (hunk 2 above)
Comment (pyrmont) [reply spec/visual/samples/racket]: Sure. Can do.

File: lib/rouge/lexers/racket.rb [hunk 3]
```diff
@@ -524,6 +525,13 @@ def self.builtins
         rule id, Name::Variable
       end
 
+      state :comment do
+        rule %r/[^|#]+/, Comment::Multiline
+        rule %r/\|#/, Comment::Multiline, :pop!
+        rule %r/#\|/, Comment::Multiline, :comment
+        rule %r/[|#]/, Comment::Multiline
```
Comment (yugenekr) [inline lib/rouge/lexers/racket.rb]: Which purpose does this line serve?

Comment (pyrmont) [reply lib/rouge/lexers/racket.rb]: This allows block comments to be nested. Otherwise the first `|#` will end the block. What's the right approach?

**Comment by yugenekr** [comment]
Hi @pyrmont, thank you for working on this.

The updates you provided look mostly fine to me (I'm not Ruby or Rouge person) in regard the particular use cases I provided.

If to dive into a bit more details, more use cases might need to be covered.

I believe [Racket parser documentation](https://docs.racket-lang.org/reference/reader.html   
) provides the list of all `#` use cases.
For example, it mentions the uppercase #T and #F also should be treated as booleans. (I'm not aware if the matching rules are case-insensitive.)

Also, see several [comments examples](http://rouge.jneen.net/v3.17.0/racket/I3xCbG9jayBjb21tZW50LgooZGVmaW5lIHggMykKaXMgbm90IGhpZ2hsaWdodGVkIGFzIHRoZSBjb21tZW50ZWQgdGV4dC4gCnwjCgojfApCbG9jayBjb21tZW50LgooZGVmaW5lIHggMykKaXMgbm90IGhpZ2hsaWdodGVkIGFzIHRoZSBjb21tZW50ZWQgdGV4dC4KfCMKCjs7ICN0cnVlIGlzIG5vdCBoaWdobGlnaHRlZCBwcm9wZXJseQooZGVmaW5lIGh0ICN0cnVlKQooZGVmaW5lIGhmICNmYWxzZSkKCjs7IE1vcmUgZXhhbXBsZXMgYmFzZWQgb24KOzsgaHR0cHM6Ly9kb2NzLnJhY2tldC1sYW5nLm9yZy9yZWZlcmVuY2UvcmVhZGVyLmh0bWwjJTI4cGFydC5fcGFyc2UtY29tbWVudCUyOQoKICN8IGEgfCMgMSA7IHJlYWRzIGVxdWFsIHRvIDEKIAogI3wgI3wgYSB8IyAxIHwjIDIgOyByZWFkcyBlcXVhbCB0byAyCiAKICM7MSAyIDsgcmVhZHMgZXF1YWwgdG8gMgogCiAjIS9iaW4vc2ggOyByZWFkcyBlcXVhbCB0byBub3RoaW5nCiAKICMhIC9iaW4vc2ggOyByZWFkcyBlcXVhbCB0byBub3RoaW5nCgoK) from this documentation page

This section might add some context to the Racket comments usage:
https://docs.racket-lang.org/style/Choosing_the_Right_Construct.html?q=comments#%28part._.Comments%29

File: lib/rouge/lexers/racket.rb (hunk 3 above)
Comment (yugenekr) [reply lib/rouge/lexers/racket.rb]: Cool, I didn't know you were adding nested comments support. I provided nested samples from the docs in the comment.

**Comment by pyrmont** [comment]
@yugenekr I think having support for commenting out s-expressions would be cool (and I don't think too difficult). Will have to add another state but should be straightforward. I'll have a tinker.

**Comment by pyrmont** [comment]
@yugenekr Famous last words, I guess. I realised looking at the list of datums in Racket that it could actually get quite complicated. I've tried a relatively simple approach that I think will cover most cases. How does it look to you?

**Comment by yugenekr** [comment]
I agree these changes should be sufficient for now. 
Other cases may be covered when needed.
Thanks!

File: lib/rouge/lexers/racket.rb [hunk 4]
```diff
@@ -511,7 +511,7 @@ def self.builtins
         rule %r/['`]#{id}/i, Str::Symbol
         rule %r/#\\([()\/'"._!\$%& ?=+-]{1}|[a-z0-9]+)/i,
           Str::Char
-        rule %r/#t(rue)?|#f(alse)?/, Name::Constant
+        rule %r/#t(rue)?|#f(alse)?/i, Name::Constant
```
Comment (yugenekr) [inline lib/rouge/lexers/racket.rb:516]: Note this would also represent #TRUE and #FALSE as valid booleans.

Comment (pyrmont) [reply lib/rouge/lexers/racket.rb:516]: That's right. I realise that might not be syntactically correct in Racket but it's a non-goal of Rouge to check syntax correctness. We will generally prefer simpler regular expressions that have no false-negatives (as it were) than more complicated expressions that are more difficult to debug and maintain.

**Comment by pyrmont** [comment]
@yugenekr Thanks for your help on this one. The update version will go out as part of the next release of Rouge. That's scheduled for Tuesday 14 April 🎉

**Comment by yugenekr** [comment]
Great! Thanks a lot reacting fast with the updates.

**Pull request merged**